rc1 update error: verify_solution

<!--
If you have a question please search or post to our Discourse site: https://discourse.julialang.org.
We use the GitHub issue tracker for bug reports and feature requests only.

If you're submitting a bug report, be sure to include as much relevant information as
possible, including a minimal reproducible example and the output of `versioninfo()`.
If you're experiencing a problem with a particular package, open an issue on that
package's repository instead.

Thanks for contributing to the Julia project!
-->

```julia
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.7.0-rc1 (2021-09-12)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

(@v1.7) pkg> up
    Updating registry at `~/.julia/registries/General.toml`
┌ Warning: gmsk[290][380][15,29] fail
└ @ Pkg.Resolve /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Pkg/src/Resolve/Resolve.jl:338
ERROR: AssertionError: verify_solution(sol, graph)
Stacktrace:
  [1] enforce_optimality!(sol::Vector{Int64}, graph::Pkg.Resolve.Graph)
    @ Pkg.Resolve /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Pkg/src/Resolve/Resolve.jl:547
  [2] resolve(graph::Pkg.Resolve.Graph)
    @ Pkg.Resolve /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Pkg/src/Resolve/Resolve.jl:87
  [3] resolve_versions!(env::Pkg.Types.EnvCache, registries::Vector{Pkg.Registry.RegistryInstance}, pkgs::Vector{Pkg.Types.PackageSpec}, julia_version::VersionNumber)
    @ Pkg.Operations /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Pkg/src/Operations.jl:336
  [4] up(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}, level::Pkg.Types.UpgradeLevel; skip_writing_project::Bool)
    @ Pkg.Operations /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Pkg/src/Operations.jl:1264
  [5] up(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; level::Pkg.Types.UpgradeLevel, mode::Pkg.Types.PackageMode, update_registry::Bool, skip_writing_project::Bool, kwargs::Base.Pairs{Symbol, Base.TTY, Tuple{Symbol}, NamedTuple{(:io,), Tuple{Base.TTY}}})
    @ Pkg.API /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Pkg/src/API.jl:333
  [6] up(pkgs::Vector{Pkg.Types.PackageSpec}; io::Base.TTY, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Pkg.API /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Pkg/src/API.jl:149
  [7] up(pkgs::Vector{Pkg.Types.PackageSpec})
    @ Pkg.API /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Pkg/src/API.jl:144
  [8] do_cmd!(command::Pkg.REPLMode.Command, repl::REPL.LineEditREPL)
    @ Pkg.REPLMode /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Pkg/src/REPLMode/REPLMode.jl:407
  [9] do_cmd(repl::REPL.LineEditREPL, input::String; do_rethrow::Bool)
    @ Pkg.REPLMode /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Pkg/src/REPLMode/REPLMode.jl:385
 [10] do_cmd
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Pkg/src/REPLMode/REPLMode.jl:376 [inlined]
 [11] (::Pkg.REPLMode.var"#24#27"{REPL.LineEditREPL, REPL.LineEdit.Prompt})(s::REPL.LineEdit.MIState, buf::IOBuffer, ok::Bool)
    @ Pkg.REPLMode /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Pkg/src/REPLMode/REPLMode.jl:549
 [12] #invokelatest#2
    @ ./essentials.jl:716 [inlined]
 [13] invokelatest
    @ ./essentials.jl:714 [inlined]
 [14] run_interface(terminal::REPL.Terminals.TextTerminal, m::REPL.LineEdit.ModalInterface, s::REPL.LineEdit.MIState)
    @ REPL.LineEdit /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/REPL/src/LineEdit.jl:2493
 [15] run_frontend(repl::REPL.LineEditREPL, backend::REPL.REPLBackendRef)
    @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/REPL/src/REPL.jl:1230
 [16] (::REPL.var"#49#54"{REPL.LineEditREPL, REPL.REPLBackendRef})()
    @ REPL ./task.jl:411
```

```julia
julia> versioninfo()
Julia Version 1.7.0-rc1
Commit 9eade6195e (2021-09-12 06:45 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.6.0)
  CPU: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 16
```

--- a/src/Resolve/Resolve.jl
+++ b/src/Resolve/Resolve.jl
@@ -255,7 +255,7 @@ function greedysolver(graph::Graph)
     for rp0 in req_inds
         # look for the highest version which satisfies the requirements
         rv0 = findlast(gconstr[rp0])
-        @assert rv0 ≠ 0 && rv0 ≠ spp[rp0]
+        @assert rv0 ≢ nothing && rv0 ≠ spp[rp0]
         sol[rp0] = rv0
         fill!(gconstr[rp0], false)
         gconstr[rp0][rv0] = true
@@ -291,14 +291,14 @@ function greedysolver(graph::Graph)
                 # or the same version was already selected, we're ok;
                 # otherwise we can't be sure what the optimal configuration is
                 # and we bail out
-                if v1 > 0 && (sol[p1] == spp[p1] || sol[p1] == v1)
-                    sol[p1] = v1
-                else
+                old_v1 = sol[p1]
+                if v1 ≡ nothing || (old_v1 ≠ v1 && old_v1 ≠ spp[p1])
                     pop_snapshot!(graph)
                     return (false, Int[])
+                elseif old_v1 == spp[p1]
+                    sol[p1] = v1
+                    push!(staged_next, p1)
                 end
-
-                p1 ∈ seen || push!(staged_next, p1)
             end
         end
         union!(seen, staged_next)
@@ -355,17 +355,18 @@ function _uninstall_unreachable!(sol::Vector{Int}, why::Vector{Union{Symbol,Int}
 
     uninst = trues(np)
     staged = Set{Int}(p0 for p0 = 1:np if !gconstr[p0][end])
-    seen = copy(staged)
+    seen = copy(staged) ∪ Set{Int}(p0 for p0 = 1:np if sol[p0] == spp[p0]) # we'll skip uninstalled packages
 
     while !isempty(staged)
         staged_next = Set{Int}()
         for p0 in staged
             s0 = sol[p0]
-            s0 == spp[p0] && continue
+            @assert s0 ≠ spp[p0]
             uninst[p0] = false
             for (j1,p1) in enumerate(gadj[p0])
+                p1 ∈ seen && continue            # we've already seen the package, or it is uninstalled
                 gmsk[p0][j1][end,s0] && continue # the package is not required by p0 at version s0
-                p1 ∈ seen || push!(staged_next, p1)
+                push!(staged_next, p1)
             end
         end
         union!(seen, staged_next)
@@ -498,8 +499,8 @@ function enforce_optimality!(sol::Vector{Int}, graph::Graph)
                             @assert lb1 ≤ s1 ≤ ub1
                             if move_up[f1]
                                 s1 > lb1 && @assert s1 == spp[f1]
-                                    # the arrangement of the range gives precedence to improving the
-                                    # current situation, but allows reinstalling a package if needed
+                                # the arrangement of the range gives precedence to improving the
+                                # current situation, but allows reinstalling a package if needed
                                 bump_range = vcat(s1:ub1, s1-1:-1:lb1)
                             else
                                 bump_range = collect(ub1:-1:lb1)

--- a/src/Resolve/Resolve.jl
+++ b/src/Resolve/Resolve.jl
@@ -425,7 +425,6 @@ function enforce_optimality!(sol::Vector{Int}, graph::Graph)
     bk_upperbound = similar(upperbound)
 
     # auxiliary sets to perform breadth-first search on the graph
-    seen = Set{Int}()
     staged = Set{Int}()
     staged_next = Set{Int}()
 
@@ -479,7 +478,6 @@ function enforce_optimality!(sol::Vector{Int}, graph::Graph)
             # if needed by another package
             try_uninstall || (lowerbound[p0] = new_s0) # note that we're in the move_up case
 
-            empty!(seen)
             empty!(staged)
             empty!(staged_next)
             push!(staged, p0)
@@ -487,12 +485,11 @@ function enforce_optimality!(sol::Vector{Int}, graph::Graph)
             while !isempty(staged)
                 for f0 in staged
                     for (j1,f1) in enumerate(gadj[f0])
-                        f1 == p0 && continue
-                        f1 ∈ seen && continue
                         s1 = sol[f1]
                         msk = gmsk[f0][j1]
-                        if try_uninstall
-                            bump_range = [s1] # when uninstalling, no further changes are allowed
+                        if f1 == p0 || try_uninstall
+                            # when uninstalling or looking at p0, no further changes are allowed
+                            bump_range = [s1]
                         else
                             lb1 = lowerbound[f1]
                             ub1 = upperbound[f1]
@@ -524,7 +521,6 @@ function enforce_optimality!(sol::Vector{Int}, graph::Graph)
                         end
                     end
                 end
-                union!(seen, staged)
                 staged, staged_next = staged_next, staged
                 empty!(staged_next)
             end

--- a/test/resolve.jl
+++ b/test/resolve.jl
@@ -396,11 +396,41 @@ end
     ]
     want_data = Dict("A"=>v"1", "B"=>v"2", "C"=>v"2", "D"=>v"2", "E"=>v"2")
     @test resolve_tst(deps_data, reqs_data, want_data)
+
+
+
+    VERBOSE && @info("SCHEME 11")
+    ## DEPENDENCY SCHEME 11: FOUR PACKAGES, WITH AN INCONSISTENCY
+    ## ref Pkg.jl issue #2740
+    deps_data = Any[
+        ["A", v"1", "C", "1"],
+        ["A", v"2", "C", "2"],
+        ["A", v"2", "D", "1"],
+        ["B", v"1", "D", "1"],
+        ["B", v"2", "D", "2"],
+        ["C", v"1", "D", "1"],
+        ["C", v"1", "B", "1"],
+        ["C", v"2", "D", "2"],
+        ["C", v"2", "B", "2"],
+        ["D", v"1"],
+        ["D", v"2"],
+    ]
+
+    @test sanity_tst(deps_data, [("A", v"2")])
+
+    # require A & B, any version (must use the highest non-inconsistent)
+    reqs_data = Any[
+        ["A", "*"],
+        ["B", "*"],
+    ]
+    want_data = Dict("A"=>v"1", "B"=>v"1", "C"=>v"1", "D"=>v"1")
+    @test resolve_tst(deps_data, reqs_data, want_data)
+
 end
 
 @testset "realistic" begin
     VERBOSE && @info("SCHEME REALISTIC")
-    ## DEPENDENCY SCHEME 11: A REALISTIC EXAMPLE
+    ## DEPENDENCY SCHEME 12: A REALISTIC EXAMPLE
     ## ref Julia issue #21485
 
     include("resolvedata1.jl")
@@ -408,7 +438,7 @@ end
     @test sanity_tst(ResolveData.deps_data, ResolveData.problematic_data)
     @test resolve_tst(ResolveData.deps_data, ResolveData.reqs_data, ResolveData.want_data)
 
-    ## DEPENDENCY SCHEME 12: A LARGER, MORE DIFFICULT REALISTIC EXAMPLE
+    ## DEPENDENCY SCHEME 13: A LARGER, MORE DIFFICULT REALISTIC EXAMPLE
     ## ref Pkg.jl issue #1949
 
     include("resolvedata2.jl")

